### ISSUE TITLE
Instance variable considered nilable despite "is_a?" check

### ISSUE THREAD

**Issue by victords** [crystal-lang/crystal#10666]
Hi, I'm not sure if this is really a bug, but considering what is described [here](https://crystal-lang.org/reference/syntax_and_semantics/if_varis_a.html) it seems to be. So, consider the code below:

```crystal
class Test
  def initialize(arg = nil)
    @arg =
      if arg.is_a?(Int32)
        arg
      else
        10
      end
  end

  def test
    puts @arg + 5
  end
end

Test.new.test
```

When trying to run it, I get the error:

```
In test.cr:12:15

 12 | puts @arg + 5
                ^
Error: undefined method '+' for Nil (compile-time type is (Int32 | Nil))
```

Shouldn't the type of `@arg` be inferred as `Int32` since both in the `if` and in the `else` branch it is assigned an `Int32`?
Curiously, if I don't include the last line (`Test.new.test`), the code compiles.

Crystal version:
```
Crystal 1.0.0 [dd40a2442] (2021-03-22)

LLVM: 10.0.0
Default target: x86_64-unknown-linux-gnu
```

Operating system: Linux Mint 20

Thanks in advance!

**Comment by Blacksmoke16** [comment]
> Curiously, if I don't include the last line (`Test.new.test`), the code compiles.

This is because code that isn't referenced isn't included in the binary.  I.e. since the `Test` class is never instantiated it's excluded from the resulting binary.

What's also worth pointing out, removing the `= nil` and passing in `nil` directly seems to make it work.  However in the meantime you could work around it by adding a `@arg : Int32` above the initializer.

**Comment by victords** [comment]
Hi, thanks for the info! I worked around it by setting the default value of the parameter to the value I was assigning in the else branch (10 in this example).

**Comment by straight-shoota** [comment]
Yeah, I think this might be really a bug. The compiler should either be able to infer the type correctly, or refuse to infer it. Not infer it incorrectly. And `typeof(if arg.is_a?(Int32) arg else 10 end)` is clearly `Int32`.

**Comment by asterite** [comment]
The reason is that, following the rules of the language, if an instance variable is assigned an argument with a default value, it takes that type too (in this case nil).

**Comment by victords** [comment]
I see. Well, even if this isn't fixed/changed, it's not a big problem since there are some simple workarounds. Thanks!

**Comment by straight-shoota** [comment]
@asterite But it's not assigned the value of the argument, it's the value of an expression which includes that argument. And the type of the expression is not nilable.

**Comment by asterite** [comment]
Right, the compiler does that: if an ivar is assigned an if, the value is guessed to be the union of the guessed branches.

The easiest fix here is to declare the instance variable type to avoid surprises.

We can try to fix it but I'm sure it will be a breaking change.

**Comment by ShaheAnsar** [comment]
Adding onto this, other nil checks fail too.
I've been trying to get the following code running
`    if @left.is_a?(BSTNode)
      @left.in_order_traversal(f)
    end
    f.call(self)
    if @right.is_a?(BSTNode)
      @right.in_order_traversal(f)
    end
`
It fails with the same error. I've tried replacing `#is_a?` with `#nil?` with inversion, as well as changing the `if` to an `unless` block with the condition uninverted.
All of them fail with - 
`Error: undefined method 'in_order_traversal' for Nil (compile-time type is (BSTNode | Nil))`

**Comment by Blacksmoke16** [comment]
@ShaheAnsar This specific case is expected.  See https://crystal-lang.org/reference/syntax_and_semantics/if_var.html#limitations.
